README.md
# php-form-encryption
This package can encrypt and decrypt forms with AES-256 by JavaScript and PHP using RSA key interchange.

SSL encryption provides a secure mechanism to protect information over public networks, but 
it is not always available: A small business private network with Wi-Fi can expose sensitive 
information, for example. There are many situations where the technical infrastructure or 
economic resources do not allow the installation of secure communication protocols.

Some times the application level encryption may be sufficient, or can even complement the 
session and/or transport level security. PHP Form Encryption  offers a application level 
encryption solution to implement in  PHP and Javascript apps development.

How does it works:

1. Server receives a client request of a web page that contains a form.

2. Server generates a session RSA key pair, and send the public key included in the html 
response.

3. Client fills out the form and generate a AES-256 key that is returned to server encrypted 
with the received public key and the AES encrypted form data. Client saves this AES key using 
browser local storage.

4. Server receives the RSA-encrypted AES key and decrypt it using the RSA private key. Then 
this AES key will be used to decrypt the received form data and to encrypt/decrypt future forms 
until it's changed or session expires.

PHP Implementation:

Session must be started before using EncryptPost class. Then, let's intercept an encrypted form:

session_start(); 

require_once './EncryptPost.class.php'; 
$crypto = new EncryptPost(1024, './openssl.cnf');

if (isset($_POST['EncryptPost'])){
	$formId = $crypto->decodeForm();
}

So, now we know the id of the submited form and $_POST superglobal contains the decrypted data. 
Before that $_POST will only contain somethig like:

var_dump($_POST) result:
array(2) {
  ["EncryptPost_key"]=>
  string(256) "5df90b95ec4fab45d50d34c917c6578f939ccbfadf9486f133850d47a3d6b2c82a277a3468ca11fc7
  b9163c385eacc2a3a4d091cf8797e55d681b0279058a9f3e334092fb03791931d22ca3847f4f9d4dec0d0a47936f01
  2b6be9723981088d0b049cff46a8e81ec93e2b4f7c3a387d36e2033754d1420a8dc800a4eec6cd0e9"
  ["EncryptPost"]=>
  string(242) "U2FsdGVkX1/53Ut6KFi36Ou/e3lIJz/5pf8FuPb1Yh//WdefKb0iyCke2/g0QPD5
BeknGV4L8dveRDbQ4kXm5YNi3nyG+/F8JWKDipA9ygHPf5KdFr6pYcfzNQjwwfd8
rIC19cl9IOJcs171tm0OBVknaloQWDwpLM/KjISdwwPiRGCtcBhkYrcdsgv6JcwD
aVuU4VunXdWJji9WAKD+1bJrThq2VLjEHhELl26y4vI="
}
Note that "EncryptPost_key" will be received only once at first time that server receives a coded 
form. Following posts will include only the "EncryptPost" value unless keys are reset.

To send the form encrypted to the server, simply include a call to the javascript crytopost class:

\<form id="form1" method="POST" action="test.php" onsubmit="return EncryptPost.encrypt('form1')">

Client may need an encrypted record to edit. Server can send it in this way:

$record = array(
	"name" => $name,
	"address" => $address,
	"zipCode" => $zip
);
$encrypted = $crypto->encodeData($record, $formId);

... and then, at the bottom of your html code include a call to javascript decrypt method:

\<script>EncryptPost.decrypt('<?php echo $encrypted;?>');\</script>


That's a simple and easy way to protect your data even with no SSL.

PHP Form Encryption requires openssl extension and PHP 5.4+

Thanks to:

- Tom Wu, author of jsbn/RSA javascript library.
- Mark Percival, author of Gibberish-AES javascript library.


changelog.txt
PHP FORM ENCRYPTION

Change log.
-------------------------------------------------------------------------------

Version 1.01    2016-09-09:
- PHP class has now more control over the RSA key pair creation.
- JavaScript class now detects RSA key change, and generates automatically a
  new AES key that's sent to the server. This means no need to manually reset
  the browser local session storage when session is destroyed and restarted, or
  when RSA keys are reset.


Version 1.00    2016-08-17:
- First public version.


test.php
<?php
/**
* Test script for EncryptPost PHP class
*/                                
session_start();
if (isset($_GET['destroySession'])){
    session_destroy();
    session_start();
}

require_once './EncryptPost.class.php';
$crypto = new EncryptPost(1024, './openssl.cnf'); // Session MUST be started.

if (isset($_GET['resetKeys'])) $crypto->reset();

// Check for FORM encrypted data
if (isset($_POST['EncryptPost'])){
    $cryptedPost = $_POST;              // Save crypted data for debug
    $formId = $crypto->decodeForm();    // Decrypt $_POST contents
    
    // Do stuff here (database record, etc). 
    // Dont forget to secure filter $_POST values.
    //
    // DON'T USE received $_POST values in the HTML code! This will transmit
    // data as clear text to the browser: Use javascript 'EncryptPost.decrypt()' 
    // method to fill your form, so data is decrypted locally at client's browser.
    if (isset($_POST['data1'])){
        $data['data1'] = filter_var($_POST['data1'], FILTER_VALIDATE_INT);
        /* ... etc ... */
    }
    
    // Encrypt processed data if you need to fill form again:
    $encrypted = $crypto->encodeData($_POST, $formId);
}
?>
<!DOCTYPE html>
<html lang="en-EN">
    <head>
        <meta charset="UTF-8">
        <title>EncryptPost</title>
        <meta name="sessionkey" content="<?php echo $_SESSION['RSA_Public_key'];?>">
        <script src="./javascript/rsa_jsbn.js"></script>
        <script src="./javascript/gibberish-aes.js"></script>
        <script src="./javascript/encryptpost.js"></script>
    </head>
    <body>

        <h1>Testing PHP Form Encryption class</h1>
        <form id="form1" method="POST" action="test.php" onsubmit="return EncryptPost.encrypt('form1')">
            Data 1: <input type="text" name="data1" value="" /><br />
            Data 2: <input type="text" name="data2" value="" /><br />
            Data 3: <input type="text" name="data3" value="" /><br />
            Data 4: <textarea cols="40" rows="5" name="data4"></textarea>
            <br />
            <input type="submit" name="submit" value="Submit" /> &nbsp;
            <input type="reset" name="reset" value="Reset" /> &nbsp;
            <a href="test.php?resetKeys=1" onclick="EncryptPost.reset();">Reset keys</a> &nbsp;
            <a href="test.php?destroySession=1">Destroy session</a>
        </form>                                           
        <!-- Fill form input fields -->
        <?php if (isset($encrypted)) { ?>
        <script>EncryptPost.decrypt('<?php echo $encrypted;?>');</script>
        <?php } ?>
        <br />
        <br />
        <?php
            // Debug
            echo '<h2>Session keys:</h2>';
            if (isset($_SESSION['RSA_Public_key'])){
                echo 'RSA public key (hex) = '. $_SESSION['RSA_Public_key'];
                echo '<br /><br />';
            }
            if (isset($_SESSION['aesKey'])){
                echo 'AES key (hex) = '. bin2hex($_SESSION['aesKey']);
                echo '<br />';
            }
            if (isset($cryptedPost)){
                echo '<h2>Received POST data:</h2><pre>';
                var_dump($cryptedPost);
                echo '</pre><br />';
                echo '<h2>Decrypted POST data:</h2><pre>';
                var_dump($_POST);
                echo '</pre><br />';
            }
        ?>
        <br /><hr />
        This class is available at <strong><a href="http://www.phpclasses.org/package/9912-PHP-Encrypt-and-decrypt-forms-with-AES-and-RSA.html">PHPClasses.org</a></strong><br />
    </body>
</html>


javascript/encryptpost.js
/**
* EncryptPost.js
* Version 1.01
* Author Francisco del Aguila <faguila@alboran.net>
* License MIT. https://opensource.org/licenses/MIT 
* 
* Javascript EncryptPost object.
* Client-side encryption/decryption for EncryptPost.class.php
* 
* Uses Gibberish-AES javascript library from Mark Percival <mark@mpercival.com>
* Uses jsbn/RSA javascript library from Tom Wu <tjw@cs.Stanford.EDU>
* 
* ver. 1.01:    Now detect changes in public RSA key and generates new AES key.
*/
var EncryptPost = {
    aeskey: null,
    aeskeyElement: null,
    separator1: " /=/ ",
    separator2: " /~/ ",
    
    init: function()
    {
        if(typeof(sessionStorage) == 'undefined'){
            alert("Error: Session storage not available.");
            return false;
        }
        
        // Get RSA public key (hex)
        element = document.querySelector('meta[name="sessionkey"]');
        if (element == null) return false;    
        RSApublic = element.getAttribute("content");
        var rsakey = new RSAKey();
        rsakey.setPublic(RSApublic, "10001");

        if (RSApublic !== sessionStorage.getItem("rsak")){
            this.reset();
            sessionStorage.setItem("rsak", RSApublic);
        }
        
        if (sessionStorage.getItem("aesk") === null){
            // Generate AES session key
            this.aeskey = this.getRandomStr(32);
            sessionStorage.setItem("aesk", this.aeskey);
            
            // Prepare input element for encrypted AES session key
            this.aeskeyElement = document.createElement('input');
            this.aeskeyElement.type = 'hidden';
            this.aeskeyElement.name = 'EncryptPost_key';
            this.aeskeyElement.value = rsakey.encrypt(this.aeskey);
        } else {
            this.aeskey = sessionStorage.getItem("aesk");
        }
        return true;
    },
    
    reset: function()
    {
        sessionStorage.removeItem("aesk");
        this.aesKey = null;
    },
    
    getRandomStr: function(length)
    {
        chars = "8<yz)AR;BOP?SQ=9T}D:N2*Zd'U5v=Jb.cef_7gEY6lmh#aC/FIu3>0Lt(wM]-V&Wq,G4+H[KXj%kin{1o$pr!sx";
        result = "";
        while (result.length < length){
            result += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        return result;
    },
    
    md5: function(str)
    {
        return GibberishAES.a2h(GibberishAES.Hash.MD5(GibberishAES.s2a(str)));
    },
    
    encrypt: function(formId)
    {
        form = document.getElementById(formId);
        if (!this.init()) return false;
        union = "";
        for (i = 0; i < form.elements.length; i++){
            if (form.elements[i].name != ''){
                union += form.elements[i].name + this.separator1;
                union += form.elements[i].value + this.separator2;
            }
            form.elements[i].name = '';
            form.elements[i].value = '';
        }
        union += 'EncryptPost_form' + this.separator1 + formId;

        if (this.aeskeyElement !== null) form.appendChild(this.aeskeyElement);
        input = document.createElement('input');
        input.type = 'hidden';
        input.name = 'EncryptPost';
        input.value = GibberishAES.enc(union, this.aeskey);
        form.appendChild(input);
        return true;
    },
    
    decrypt: function(data)
    {
        this.aeskey = sessionStorage.getItem("aesk");
        if (this.aeskey === null || data === null) return false;
        
        decrypted = GibberishAES.dec(data, this.aeskey);
        var key = [];
        var form = null;

        pairs = decrypted.split(this.separator2);
        for (i=0; i<pairs.length; i++){
            pair = pairs[i].split(this.separator1);
            if (pair[0] == 'EncryptPost_form'){
                form = document.getElementById(pair[1]);    
            } else {
                if (pair[0] != '') key[pair[0]] = pair[1];
            }
        }
        
        if (form == null) return false;
        
        for (i = 0; i < form.elements.length; i++){
            form.elements[i].value = key[form.elements[i].name];        
        }
        
        return true;
    }
}

EncryptPost.class.php
<?php
if (!extension_loaded('openssl')) die('ERROR: Openssl PHP extension not loaded.'); 

/**
* @name EncryptPost.class.php
* @author Francisco del Aguila <faguila@alboran.net>
* @version 1.01
* @license MIT https://opensource.org/licenses/MIT
* 
* Requires PHP 5.4+
* Requires openssl extension
* 
* V. 1.01:  - If no RSA private key is present in session when EncryptPost_key is
*             received, the script will exit with error.
*           - Detect failure in new RSA key creation.
*       
*/

class EncryptPost
{   
    const SEPARATOR1 = ' /=/ ';
    const SEPARATOR2 = ' /~/ ';
    
    protected $opensslCnf; // Filename of the openssl.cnf config file.
    protected $RSAkeyLength;
    
    public function __construct($RSAkeyLength = 1024, $opensslCnf = './openssl.cnf')
    {
        // Create the session RSA key pair if doesn't exists.
        $this->RSAkeyLength = $RSAkeyLength;
        $this->opensslCnf = $opensslCnf;
        if (!isset($_SESSION['RSA_Private_key'])) $this->createSessionKeys();
    }
    
    /**
    * Create new RSA public and private session keys
    * 
    * @return array
    */
    public function createKeys()
    {
        $config = array(
            "config" => $this->opensslCnf,
            "private_key_bits" => $this->RSAkeyLength,
            "private_key_type" => OPENSSL_KEYTYPE_RSA
        );
        
        $privatekey = '';
        $keydetails = false;
        
        $pkey = openssl_pkey_new($config);
        if ($pkey) $pkeyCreated = openssl_pkey_export($pkey, $privatekey, null, $config);
        if ($pkeyCreated) $keydetails = openssl_pkey_get_details($pkey);
        
        if (!$keydetails) die("RSA key creation failed");
        
        return array(
            "rsaPrivateKey" => $privatekey,
            "rsaPublicKeyHex" => bin2hex($keydetails['rsa']['n'])
        );
    }
    
    /**
    * Set session RSA keys
    * 
    * @param int $keyLength
    */
    public function createSessionKeys()
    {
        $keyArray = $this->createKeys();
        $_SESSION['RSA_Private_key'] = $keyArray['rsaPrivateKey'];
        $_SESSION['RSA_Public_key'] = $keyArray['rsaPublicKeyHex'];
    }
    
    /**
    * Generates a gibberish-aes compatible hash whith 32 bytes key and iv.
    * $salt length must be 8 bytes.
    * 
    * @param string $salt
    * @param string $key
    * 
    * @return string
    */
    protected function getHash($salt, $key)
    {
        $hash1 = md5($key . $salt, true);
        $hash2 = md5($hash1 . $key . $salt, true);
        $hash3 = md5($hash2 . $key . $salt, true);
        $hash4 = $hash1 . $hash2 . $hash3;

        $hash['key'] = substr($hash4, 0, 32);
        $hash['iv'] = substr($hash4, 32, 16);
        
        return $hash;
    }
    
    /**
    * Encrypt a string using AES 256 bits cbc mode.
    * 
    * @param string $decrypted
    * @param string $key
    * 
    * @return string base64 encoded
    */
    public function aes256cbc_encrypt($decrypted, $key)
    {
        $salt = openssl_random_pseudo_bytes(8);
        $hash = $this->getHash($salt, $key);
        $encrypted = openssl_encrypt($decrypted, "aes-256-cbc", $hash['key'], true, $hash['iv']);

        return base64_encode('Salted__' . $salt . $encrypted);
    }
    
    /**
    * Decrypt a string with AES 256 bits cbc.
    * 
    * @param string $encrypted Base64 encoded
    * @param string $key
    * 
    * @return string 
    */
    public function aes256cbc_decrypt($encrypted, $key)
    {
        $encrypted = base64_decode($encrypted);
        $salt = substr($encrypted, 8, 8);
        $encrypted = substr($encrypted, 16);
        $hash = $this->getHash($salt, $key);
        return openssl_decrypt($encrypted, "aes-256-cbc", $hash['key'], true, $hash['iv']);
    }
    
    /**
    * Delete session AES key and regenerate RSA keys
    */
    public function reset()
    {
        unset($_SESSION['aesKey']);
        $this->createSessionKeys();    
    }
    
    /**
    * Decode a submitted FORM ($_POST)
    * 
    * $_POST must contain two keys when first time executed in a session:
    * 'EncryptPost' : With the AES 256 encrypted serialized form data.
    * 'EncryptPost_key': Whith the RSA encrypted (with public key) AES 256 key.
    * 
    * Next times in that session 'EncryptPost_key' is not needed.
    * 
    * The form id must be serialized inside the 'EncryptPost' content assigned to
    * the 'EncryptPost_form' key.
    * 
    * The decrypted values are assigned to the $_POST superglobal.
    * 
    * @return string Form id
    */
    public function decodeForm()
    {   
        if (!isset($_POST['EncryptPost'])) return false; // Nothing to decrypt
        
        // Get and decrypt EncryptPost_key if present
        if (isset($_POST['EncryptPost_key'])){
            if (!isset($_SESSION['RSA_Private_key'])) die("RSA key not found");
            $rsaPrivateKey = openssl_pkey_get_private($_SESSION['RSA_Private_key']);
            $encrypted = pack('H*', $_POST['EncryptPost_key']);
            $aesKey = '';
            if (!openssl_private_decrypt($encrypted, $aesKey, $rsaPrivateKey)) return false;
            $_SESSION['aesKey'] = $aesKey;
            unset($_POST['EncryptPost_key']);
        }
        
        if (!isset($_SESSION['aesKey'])) die("Decrypt: AES key not found");
        
        $aesKey = $_SESSION['aesKey'];

        // Decrypt post
        $encrypted = $_POST['EncryptPost'];
        $decrypted = $this->aes256cbc_decrypt($encrypted, $aesKey);
        
        $pairs = explode(self::SEPARATOR2, $decrypted);
        foreach ($pairs as $pair){
            list($key, $value) = explode(self::SEPARATOR1, $pair);
            $_POST[$key] = $value;
        }
        
        $formId = $_POST['EncryptPost_form'];
        
        unset($_POST['EncryptPost'], $_POST['EncryptPost_form']);
        
        return $formId;
    }
    
    /**
    * Encrypt the $data key/values array of formId
    * 
    * @param array $data
    * @param string $form Form id.
    * 
    * @return array
    */
    public function encodeData($data, $formId)
    {
        if (!$data || !$formId) return false;
        
        if (!isset($_SESSION['aesKey'])) die("Encrypt: AES key not found");
        
        // Serialize data and form id
        $union = '';
        foreach ($data as $key=>$value){
            $union .= $key . self::SEPARATOR1 . $value . self::SEPARATOR2;    
        }
        $union .= 'EncryptPost_form' . self::SEPARATOR1 . $formId;
        
        // Encrypt the serialized data
        $encrypted = $this->aes256cbc_encrypt($union, $_SESSION['aesKey']);
        
        // Return array of encripted key and data with formId
        return $encrypted;
    }    
}
?>